src/DiskIO/IpcIo/IpcIoFile.cc

```diff
@@ -447,23 +444,57 @@ IpcIoFile::canWait() const
     return false; // do not want to wait that long
 }
 
-/// called when coordinator responds to worker open request
 void
-IpcIoFile::HandleOpenResponse(const Ipc::StrandMessage &response)
+IpcIoFile::HandleStrandReadyResponse(const Ipc::StrandReady &response)
 {
-    debugs(47, 7, HERE << "coordinator response to open request");
-    for (IpcIoFileList::iterator i = WaitingForOpen.begin();
-            i != WaitingForOpen.end(); ++i) {
-        if (response.strand.tag == (*i)->dbName) {
-            (*i)->openCompleted(&response);
-            WaitingForOpen.erase(i);
-            return;
-        }
+    debugs(47, 7, "disker" << response.strand.kidId <<  " " << response.strand.tag << " is ready");
+    if (const auto file = StopWaiting(response.strand))
+        file->openCompleted(&response);
+    else
+        debugs(47, 4, "premature or late disker response");
+}
+
+/// reschedules IpcIoFile open request timeout for the head of the waiting queue
+/// \param file an IpcIoFile to be queued
+void
+IpcIoFile::StartWaiting(const Pointer &file)
+{
+    if (file) {
+        auto expireTime = current_time;
+        expireTime.tv_sec += Timeout;
+        WaitingForOpen.emplace(expireTime, file);
+    }
+
+    assert(!WaitingForOpen.empty());
+    eventDelete(&IpcIoFile::OpenTimeout, nullptr); // deletes all OpenTimeout handlers
+    const auto interval = tvSubDsec(current_time, WaitingForOpen.begin()->first);
+    eventAdd("IpcIoFile::OpenTimeout", &IpcIoFile::OpenTimeout, nullptr, interval, 0, false);
+}
+
+/// removes the IpcIoFile specified by the response from the waiting queue
+IpcIoFile::Pointer
+IpcIoFile::StopWaiting(const Ipc::StrandCoord &responseStrand)
+{
+    const auto it = std::find_if(WaitingForOpen.begin(), WaitingForOpen.end(),
+    [&responseStrand](const FileWait &pair) { return pair.second->dbName == responseStrand.tag; });
+    if (it != WaitingForOpen.end()) {
+        const auto file = it->second;
+        WaitingForOpen.erase(it);
```

Comment:
If possible, please delete the IpcIoFile::OpenTimeout event when the queue becomes empty.